File: train-bus.py
# -*- coding: utf-8 -*-
"""
Created on Sun Apr 29 04:29:01 2024

@author: Karim Itani
"""
train_prices = [8, 6, 5, 800]
bus_prices = [8, 12, 5, 100]

n = len(train_prices) 
switch_cost = 10

dp = [0] * n
cost = 0
on_train = False  # boolean used for logic control flow

if train_prices[0] < bus_prices[0]:
    
    cost = train_prices[0]
    on_train = True

elif train_prices[0] > bus_prices[0]:

    cost = bus_prices[0]
    on_train = False


else:  # train_prices[0] == bus_prices[0]
    
    if train_prices[1]  > bus_prices[1]:
        cost = bus_prices[0]  # choose bus 
        on_train = False
    else:
        cost = train_prices[0]  # choose train 
        on_train = True

dp[0] = cost

for i in range(1, n):

    if on_train:
        if train_prices[i] > bus_prices[i] + switch_cost:

            cost += switch_cost + bus_prices[i]
            on_train = False
        else:

            cost += train_prices[i]
    else:

        if bus_prices[i] > train_prices[i] + switch_cost:

            cost += switch_cost + train_prices[i]
            on_train = True
        else:

            cost += bus_prices[i]

    dp[i] = cost

print(dp)
print("Suzie Q can reach her final destination by paying "+ str(dp[-1]) + " dollars")

File: Train_Bus.java
package cp3;

public class Train_Bus {
	public static void main(String[] args) {
        int[] trainPrices = {5, 100, 5, 100, 5};
        int[] busPrices = {100, 5, 100, 5, 100};
        int n = trainPrices.length;  
        int switchCost = 10;
        
        int[] dp = new int[n];
        int cost = 0;
        boolean onTrain = false;  // Boolean used for logic control flow
        
        if(trainPrices[0] < busPrices[0]){
            cost = trainPrices[0];
            onTrain = true;
        } 
        else if(trainPrices[0] > busPrices[0]){
            cost = busPrices[0];
            onTrain = false;
        } 
        
        else{  // trainPrices[0] == busPrices[0]
            if (trainPrices[1] > busPrices[1]){
                cost = busPrices[0];  // choose bus
                onTrain = false;
            } 
            else{
                cost = trainPrices[0];  // choose train
                onTrain = true;
            }
        }
        
        dp[0] = cost;
        
        // Process the rest of the stations
        for(int i = 1; i < n; i++){
            if(onTrain){
            	
                if(trainPrices[i] > busPrices[i] + switchCost){
                    cost += switchCost + busPrices[i];
                    onTrain = false;
                } 
                else{
                    cost += trainPrices[i];
                }
            } 
            else{
                
            	if (busPrices[i] > trainPrices[i] + switchCost){
                    cost += switchCost + trainPrices[i];
                    onTrain = true;
                } 
            	
            	else{
                    cost += busPrices[i];
                }
            }
            
            dp[i] = cost;
        }
        
        for (int i = 0; i < n; i++){
            int res = dp[i];
        	System.out.print(res + " ");
        }
        System.out.println();

    	System.out.println("Suzie Q can reach her final destination by paying " + dp[n-1]+ " dollars!");

    }

}
